GracefulKiller/GracefulKiller.py

```diff
@@ -14,13 +15,59 @@ def __len__(self):
 
 # process SIGTERM and SIGINT signals gracefully
 class GracefulKiller:
-    def __init__(self):
+
+    def __new__(cls, *args, **kwargs):
+        if not hasattr(cls, 'instance'):
+            cls.instance = super(GracefulKiller, cls).__new__(cls)
+        return cls.instance
+
+    def __init__(self, shutdown_handler=None):
+        assert shutdown_handler is None or callable(shutdown_handler)
+        self.shutdown_handler = shutdown_handler
         self.kill_now = KillerEvent()
         for sig in ('TERM', 'HUP', 'INT'):
             signalnum = getattr(signal, 'SIG' + sig, None)
             if signalnum is None:
                 continue
             signal.signal(signalnum, self.exit_gracefully)
 
-    def exit_gracefully(self, signum, frame):
+    def kill(self):
         self.kill_now.set()
+
+    def exit_gracefully(self, signum, frame):
+        self.kill()
+
+    def shutdown(self):
+        if self.shutdown_handler:
+            self.shutdown_handler()
+
+
+class Loop:
+
+    def __new__(cls, *args, **kwargs):
+        if not hasattr(cls, 'instance'):
+            cls.instance = super(Loop, cls).__new__(cls)
+        return cls.instance
+
+    def __init__(self, killer, delay):
+        assert killer is not None and isinstance(killer, GracefulKiller)
```

Comment:
why it can't be just like this:
`assert killer and isinstance(killer, GracefulKiller)`?